# Предобработка датасета

## Обзор данных

In [99]:
import pandas as pd
import numpy as np
import ast

In [100]:
df = pd.read_csv('games.csv')

In [101]:
df.sample()

Unnamed: 0                Title  Release Date   
462         462  The Legend of Zelda  Feb 21, 1986  \

                             Team  Rating Times Listed Number of Reviews   
462  ['Nintendo EAD', 'Nintendo']     3.2         1.2K              1.2K  \

            Genres                                            Summary   
462  ['Adventure']  The Legend of Zelda is the first title in the ...  \

                                               Reviews Plays Playing Backlogs   
462  ['Managed to 100% this in less than two hours,...  9.3K     132     1.5K  \

    Wishlist  
462      463

In [102]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1512 entries, 0 to 1511
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         1512 non-null   int64  
 1   Title              1512 non-null   object 
 2   Release Date       1512 non-null   object 
 3   Team               1511 non-null   object 
 4   Rating             1499 non-null   float64
 5   Times Listed       1512 non-null   object 
 6   Number of Reviews  1512 non-null   object 
 7   Genres             1512 non-null   object 
 8   Summary            1511 non-null   object 
 9   Reviews            1512 non-null   object 
 10  Plays              1512 non-null   object 
 11  Playing            1512 non-null   object 
 12  Backlogs           1512 non-null   object 
 13  Wishlist           1512 non-null   object 
dtypes: float64(1), int64(1), object(12)
memory usage: 165.5+ KB


## Пропуски

In [103]:
df.isna().sum()

Unnamed: 0            0
Title                 0
Release Date          0
Team                  1
Rating               13
Times Listed          0
Number of Reviews     0
Genres                0
Summary               1
Reviews               0
Plays                 0
Playing               0
Backlogs              0
Wishlist              0
dtype: int64

In [104]:
df[df['Rating'].isna()==1].T

587    
Unnamed: 0                                                       587  \
Title                                              Final Fantasy XVI   
Release Date                                            Jun 22, 2023   
Team               ['Square Enix', 'Square Enix Creative Business...   
Rating                                                           NaN   
Times Listed                                                     422   
Number of Reviews                                                422   
Genres                                                       ['RPG']   
Summary            Final Fantasy XVI is an upcoming action role-p...   
Reviews                                                           []   
Plays                                                             37   
Playing                                                           10   
Backlogs                                                         732   
Wishlist                                                        2.4K   

                                       649    
Unnamed: 0                              649  \
Title                     Death Stranding 2   
Release Date                releases on TBD   
Team                 ['Kojima Productions']   
Rating                                  NaN   
Times Listed                            105   
Number of Reviews                       105   
Genres             ['Adventure', 'Shooter']   
Summary                                 NaN   
Reviews                                  []   
Plays                                     3   
Playing                                   0   
Backlogs                                209   
Wishlist                                644   

                                                                713    
Unnamed: 0                                                       713  \
Title                                      Final Fantasy VII Rebirth   
Release Date                                            Dec 31, 2023   
Team                                                 ['Square Enix']   
Rating                                                           NaN   
Times Listed                                                     192   
Number of Reviews                                                192   
Genres                                                            []   
Summary            This next standalone chapter in the FINAL FANT...   
Reviews                                                           []   
Plays                                                             20   
Playing                                                            3   
Backlogs                                                         354   
Wishlist                                                        1.1K   

                                                                719    
Unnamed: 0                                                       719  \
Title                                                      Lies of P   
Release Date                                            Aug 01, 2023   
Team                                     ['NEOWIZ', 'Round8 Studio']   
Rating                                                           NaN   
Times Listed                                                     175   
Number of Reviews                                                175   
Genres                                                       ['RPG']   
Summary            Inspired by the familiar story of Pinocchio, L...   
Reviews                                                           []   
Plays                                                              5   
Playing                                                            0   
Backlogs                                                         260   
Wishlist                                                         939   

                                                                726    
Unnamed: 0                                 

Есть игры, которые ещё только будут выходить в этом и следующем годах, поэтому их рейтинг пока неизвестен. Заменю пропуск на -1.

In [105]:
df[df['Team'].isna()==1]

Unnamed: 0                 Title  Release Date Team  Rating   
1245        1245  NEET Girl Date Night  Oct 21, 2022  NaN     2.7  \

     Times Listed Number of Reviews            Genres   
1245           21                21  ['Visual Novel']  \

                                                Summary   
1245  Your friend sets you up on a date with his NEE...  \

                                                Reviews Plays Playing   
1245  ['this sucked. "Omg she is literally me" is no...   106       1  \

     Backlogs Wishlist  
1245       44       42

Как подсказывает гугл разработчик этой игры - Hitsuji Goods. Поэтому внесём это в наши данные

In [106]:
df.loc[df['Title']=='NEET Girl Date Night','Team'] = "['Hitsuji Goods']"
df['Rating'] = df['Rating'].fillna(-1)

## Обработка даты релиза

In [107]:
df[df['Release Date'] == 'releases on TBD']

Unnamed: 0                              Title     Release Date   
644          644                          Deltarune  releases on TBD  \
649          649                  Death Stranding 2  releases on TBD   
1252        1252  Elden Ring: Shadow of the Erdtree  releases on TBD   

                                                Team  Rating Times Listed   
644                                      ['tobyfox']     4.3          313  \
649                           ['Kojima Productions']    -1.0          105   
1252  ['FromSoftware', 'Bandai Namco Entertainment']     4.8           18   

     Number of Reviews                                            Genres   
644                313  ['Adventure', 'Indie', 'Music', 'Puzzle', 'RPG']  \
649                105                          ['Adventure', 'Shooter']   
1252                18                              ['Adventure', 'RPG']   

                                                Summary   
644   UNDERTALE's parallel story, DELTARUNE. Meet ne...  \
649                                                 NaN   
1252  An expansion to Elden Ring setting players on ...   

                                                Reviews Plays Playing   
644   ['Spamton is so hot, I want to kiss him in the...  1.3K      83  \
649                                                  []     3       0   
1252  ['I really loved that they integrated Family G...     1       0   

     Backlogs Wishlist  
644       468      617  
649       209      644  
1252       39      146

У 3-х игр дата релиза неизвестна. Погуглив выяснил, что Death Stranding 2 и Elden Ring: Shadow of the Erdtree ещё не вышли. А Deltarune вышел на хеллуин 31 октября 2018 года, поэтому невышедшие удалю, а в Deltarune поставлю корректную дату.

In [108]:
df.loc[644,'Release Date'] = 'Oct 31, 2018'
df = df.drop([649,1252]).drop(columns='Unnamed: 0').reset_index(drop=True)
df['Release Date'] = pd.to_datetime(df['Release Date'])

Переведу числа из текстового формата K (тысячи) в целочисленный формат.

In [109]:
k_cols = ['Times Listed',
          'Number of Reviews',
          'Plays', 
          'Playing',
          'Backlogs', 
          'Wishlist']

for col in k_cols:
    df[col] = (df[col].str.replace('K','').astype(float)*1000).astype(int)

In [110]:
df.head(1).T

0
Title                                                     Elden Ring
Release Date                                     2022-02-25 00:00:00
Team                  ['Bandai Namco Entertainment', 'FromSoftware']
Rating                                                           4.5
Times Listed                                                    3900
Number of Reviews                                               3900
Genres                                          ['Adventure', 'RPG']
Summary            Elden Ring is a fantasy, action and open world...
Reviews            ["The first playthrough of elden ring is one o...
Plays                                                          17000
Playing                                                         3800
Backlogs                                                        4600
Wishlist                                                        4800

## Удаление дубликатов

In [111]:
df.duplicated().sum()

382

382 Полных дубликата, так как они не предоставляют ценности, то удалю их.

In [112]:
df = df.drop_duplicates().reset_index(drop=True)

## Обработка жанров и команд

In [113]:
df['Team'] = df['Team'].apply(lambda x: ast.literal_eval(x))
team = pd.DataFrame({
    'Title': df['Title'].tolist(),
    'Team': df['Team'].tolist() }).explode('Team')

team.to_csv('teams.csv')

team

Title   
0                                  Elden Ring  \
0                                  Elden Ring   
1                                       Hades   
2     The Legend of Zelda: Breath of the Wild   
2     The Legend of Zelda: Breath of the Wild   
...                                       ...   
1124                        Team Sonic Racing   
1125                           Dragon's Dogma   
1126                          Baldur's Gate 3   
1127                 The LEGO Movie Videogame   
1127                 The LEGO Movie Videogame   

                                     Team  
0              Bandai Namco Entertainment  
0                            FromSoftware  
1                        Supergiant Games  
2                                Nintendo  
2     Nintendo EPD Production Group No. 3  
...                                   ...  
1124                                 Sega  
1125                               Capcom  
1126                       Larian Studios  
1127                             WB Games  
1127                            TT Fusion  

[2020 rows x 2 columns]

In [120]:
print('Топ 10 команд по количеству выпущенных игр')
team['Team'].value_counts().head(10)

Топ 10 команд по количеству выпущенных игр


Team
Nintendo                      174
Capcom                         64
Square Enix                    59
Sega                           48
Electronic Arts                43
Konami                         31
Activision                     30
Bandai Namco Entertainment     29
Ubisoft Entertainment          28
Atlus                          26
Name: count, dtype: int64

In [115]:
df['Genres'] = df['Genres'].apply(lambda x: ast.literal_eval(x))
genres = pd.DataFrame({
    'Title': df['Title'].tolist(),
    'Genres': df['Genres'].tolist() }).explode('Genres')

genres.to_csv('genres.csv')

genres

Title               Genres
0                   Elden Ring            Adventure
0                   Elden Ring                  RPG
1                        Hades            Adventure
1                        Hades              Brawler
1                        Hades                Indie
...                        ...                  ...
1126           Baldur's Gate 3             Strategy
1126           Baldur's Gate 3             Tactical
1126           Baldur's Gate 3  Turn Based Strategy
1127  The LEGO Movie Videogame            Adventure
1127  The LEGO Movie Videogame               Puzzle

[2713 rows x 2 columns]

In [116]:
genres['Genres'].value_counts()

Genres
Adventure              732
RPG                    374
Shooter                264
Platform               243
Indie                  207
Puzzle                 133
Strategy               117
Brawler                109
Simulator               95
Turn Based Strategy     71
Fighting                64
Arcade                  61
Visual Novel            54
Tactical                42
Racing                  37
Point-and-Click         36
Music                   22
Sport                   20
Card & Board Game       13
Real Time Strategy       9
MOBA                     3
Quiz/Trivia              3
Pinball                  1
Name: count, dtype: int64

Сохраню предобработанный датасет

In [121]:
df.to_csv('games_preprocessed')

## Итоги

- Провёл обзор данных.
- Удалил дубликаты.
- Заполнил пропуски.
- Обработал дату релиза.
- Перевёл численные значения из строкового формата *K (тыс.) в целочисленный формат *000. (К примеру, было 18.5K, стало 18500)
- Так как одна и та же игра может относиться к нескольким жанрам или командам разработчиков, то я выделил ещё два датасета "teams.csv" и "genres.csv".

Данные готовы к визуализации, которую буду делать в Yandex DataLens.